File: MCP/AID-tool-gateway/scripts/gateway-call.mjs
#!/usr/bin/env node
// Purpose: call one gateway tool over stdio and print structured JSON output.
// Usage: node scripts/gateway-call.mjs --tool tool_search --payload '{"query":"..."}'

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const repoRoot = path.resolve(__dirname, "..", "..", "..");
const gatewayLauncher = path.join(repoRoot, "MCP", "run-tool-gateway.sh");
const SHUTDOWN_TIMEOUT_MS = 2000;

function parseArgs(argv) {
  let toolName = "";
  let payloadRaw = "{}";
  for (let index = 2; index < argv.length; index += 1) {
    const token = argv[index];
    if (token === "--tool" && index + 1 < argv.length) {
      toolName = argv[index + 1];
      index += 1;
      continue;
    }
    if (token === "--payload" && index + 1 < argv.length) {
      payloadRaw = argv[index + 1];
      index += 1;
      continue;
    }
  }

  if (!toolName.trim()) {
    throw new Error("Missing required --tool argument.");
  }

  let payload = {};
  try {
    payload = JSON.parse(payloadRaw);
  } catch (error) {
    throw new Error(`Invalid JSON passed to --payload: ${error instanceof Error ? error.message : String(error)}`);
  }
  if (payload === null || typeof payload !== "object" || Array.isArray(payload)) {
    throw new Error("--payload must be a JSON object.");
  }
  return { toolName: toolName.trim(), payload };
}

async function main() {
  const { toolName, payload } = parseArgs(process.argv);
  const transport = new StdioClientTransport({ command: gatewayLauncher });
  const client = new Client({ name: "gateway-call", version: "0.0.1" });

  try {
    await client.connect(transport);
    const result = await client.callTool({ name: toolName, arguments: payload });
    process.stdout.write(
      `${JSON.stringify(
        {
          tool: toolName,
          isError: Boolean(result?.isError),
          structuredContent: result?.structuredContent ?? null,
          content: Array.isArray(result?.content) ? result.content : [],
        },
        null,
        2
      )}\n`
    );
  } finally {
    await closeWithTimeout(() => client.close(), SHUTDOWN_TIMEOUT_MS);
    await closeWithTimeout(() => transport.close(), SHUTDOWN_TIMEOUT_MS);
  }
}

async function closeWithTimeout(closeFn, timeoutMs) {
  await Promise.race([
    Promise.resolve()
      .then(() => closeFn())
      .catch(() => {}),
    new Promise((resolve) => {
      const timer = setTimeout(resolve, timeoutMs);
      if (typeof timer.unref === "function") {
        timer.unref();
      }
    }),
  ]);
}

main()
  .then(() => {
    process.exit(0);
  })
  .catch((error) => {
    process.stderr.write(`${error instanceof Error ? error.stack ?? error.message : String(error)}\n`);
    process.exit(1);
  });

File: AIDZeroCode/MCP/tool-gateway/scripts/gateway-call.py
#!/usr/bin/env python3
"""Call one gateway tool over stdio and print JSON output."""

from __future__ import annotations

import argparse
import asyncio
import json
from pathlib import Path
from typing import Any

from mcp import ClientSession
from mcp.client.stdio import StdioServerParameters, stdio_client


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Call MCP tool gateway")
    parser.add_argument("--tool", required=True)
    parser.add_argument("--payload", default="{}")
    return parser.parse_args()


async def run(tool_name: str, payload: dict[str, Any]) -> dict[str, Any]:
    repo_root = Path(__file__).resolve().parents[3]
    gateway_launcher = repo_root / "MCP" / "run-tool-gateway.sh"

    params = StdioServerParameters(command=str(gateway_launcher), args=[])
    async with stdio_client(params) as (read_stream, write_stream):
        async with ClientSession(read_stream, write_stream) as session:
            await session.initialize()
            result = await session.call_tool(tool_name, payload)

    content = []
    for block in result.content:
        content.append(block.model_dump(exclude_none=True))

    output: dict[str, Any] = {
        "tool": tool_name,
        "isError": bool(result.isError),
        "structuredContent": result.structuredContent,
        "content": content,
    }

    if output["structuredContent"] is None:
        parsed = _extract_json_from_text_content(content)
        if isinstance(parsed, dict):
            output["structuredContent"] = parsed
    return output


def _extract_json_from_text_content(content: list[dict[str, Any]]) -> dict[str, Any] | None:
    if len(content) != 1:
        return None
    block = content[0]
    if block.get("type") != "text" or not isinstance(block.get("text"), str):
        return None
    raw = block["text"].strip()
    if not raw.startswith("{"):
        return None
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError:
        return None
    return parsed if isinstance(parsed, dict) else None


def main() -> None:
    args = parse_args()
    try:
        payload = json.loads(args.payload)
    except json.JSONDecodeError as error:
        raise SystemExit(f"Invalid JSON passed to --payload: {error}") from error
    if not isinstance(payload, dict):
        raise SystemExit("--payload must be a JSON object")

    result = asyncio.run(run(args.tool.strip(), payload))
    print(json.dumps(result, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
